src/lib/structures/uid.ml

```diff
@@ -26,46 +26,83 @@
 (**************************************************************************)
 
 module DStd = Dolmen.Std
-module DE = DStd.Expr
+module DE = Dolmen.Std.Expr
 
-type t =
-  | Hstring : Hstring.t -> t
-  | Dolmen : 'a DE.id -> t
+type _ t =
+  | Hstring : Hstring.t -> 'a t
+  | Term_cst : DE.term_cst -> DE.term_cst t
+  | Ty_cst : DE.ty_cst -> DE.ty_cst t
+  | Ty_var : DE.ty_var -> DE.ty_var t
 
-let[@inline always] of_dolmen id = Dolmen id
+type term_cst = DE.term_cst t
+type ty_cst = DE.ty_cst t
+type ty_var = DE.ty_var t
+
+let order_tag : int DStd.Tag.t = DStd.Tag.create ()
+
+let has_order id =
```

Comment:
It is somewhat confusing that `has_order` can return `true` for terms that have no order attached. Something like `has_order_if_constructor` would be clearer.